articles/modules/ROOT/pages/how-to-force-direction-of-expansion.adoc

```diff
@@ -0,0 +1,128 @@
+= How to force direction of an expansion
+:slug: how-to-force-direction-of-expansion
+:author: Andrew Bowman
+:neo4j-versions: 3.5, 4.0, 4.1, 4.2, 4.3, 4.4, 5.x
+:tags: cypher
+:category: cypher
+
+Like SQL, Cypher is a declarative query language.
+This is most evident with Match patterns, which describe what you want to find in the graph.
+You do not dictate to it how to find these patterns, the way you might in an imperative programming language.
+
+For the most part, this works out well, with the planner using the counts and statistical metadata in the graph to figure out how to find the patterns you want to match against, and which nodes to lookup (via indexes) to serve as anchor nodes that you will begin your expansions from.
+
+However, sometimes the counts and metadata information is not sufficient for producing an optimal plan, particularly when it comes to the direction of certain expansions.
+That is, for some expansions, traversing in one direction may be considerably cheap, but expanding in the opposite direction may be very expensive.
+
+This article discusses the problem, and provides means by which you can force the planner to expand in a certain direction if needed.
+
+== When one direction is clearly more expensive
+
+Consider if you have a graph of social graph contact data, which may include this pattern for all contacts:
+
+[source,cypher]
+----
+(:User)-[:LIVES_IN]->(:Country)
+----
+
+If we are querying for a specific user in a specific country, we might have part of the query include:
+
+[source,cypher]
+----
+MATCH (user:User {id:12345})-[:LIVES_IN]->(country:Country {name:'United States'})
+----
+
+There are several different ways we can imagine how this match portion might get planned, subject to indexes present and graph metadata.
+
+If only one of these nodes is chosen as an anchor node, the expansion will start there, expand the relationship to the other node, and then filter the other node's label and properties to find the matches.
+
+If both nodes are looked up via index, still only one will be the anchor node that we expand from, so the expansion process should remain about the same, but filtering will be more efficient, only having to filter on the node's internal graph id to see if the other node is the same one we matched earlier.
```

Comment:
I would probably rewrite this to something along the lines of:

If both nodes are looked up from an index, an "expand-into" operation is happening.
Expansion will be more efficient if the node with the smaller degree is chosen. 

(At runtime the degree of both nodes will be taken into account)

I guess the trickier case if it only picks one index and the wrong side (dense node).
Then you need to force it with using index or using join on (the other node)